subsys/bluetooth/host/hci_core.c

```diff
@@ -1779,7 +1780,14 @@ int bt_le_set_data_len(struct bt_conn *conn, uint16_t tx_octets, uint16_t tx_tim
 	cp->tx_octets = sys_cpu_to_le16(tx_octets);
 	cp->tx_time = sys_cpu_to_le16(tx_time);
 
-	return bt_hci_cmd_send(BT_HCI_OP_LE_SET_DATA_LEN, buf);
+	err = bt_hci_cmd_send(BT_HCI_OP_LE_SET_DATA_LEN, buf);
+	if (err) {
+		return err;
+	}
+
+	conn->le.data_len.tx_max_len = tx_octets;
+	conn->le.data_len.tx_max_time = tx_time;
```

Comment:
Are you sure this is correct? Notice that we're using `bt_hci_cmd_send()` and *not* `bt_hci_cmd_send_sync()` which means that the `err` doesn't tell you if the command failed or not (in fact, the command hasn't even completed at this point). It seems to me like the updating should hapen when the command completes, so if possible, change to using the `_sync()` variant.